src/snowflake/connector/storage_client.py

```diff
@@ -445,9 +445,10 @@ def _abort_multipart_upload(self) -> None:
     def delete_client_data(self) -> None:
         """Deletes the tmp_dir and closes the source stream belonging to this client.
         This function is idempotent."""
-        if os.path.exists(self.tmp_dir):
-            logger.debug(f"cleaning up tmp dir: {self.tmp_dir}")
-            shutil.rmtree(self.tmp_dir)
+        with self.lock:
```

Comment:
The stackoverflow answer claims that `An object deconstructor may not always have a predictable behaviour` but for CPython, PEP-442 mandates that an object’s finalizer should always called exactly once. The stackoverflow answer states that their race condition comes from using `multiprocessing` where two processes try to delete the same directory, but that shouldn't apply to us. I'm curious as to where the race condition is coming from here.

Reply:
possibly from this function: _upload_files_in_parallel? In my case, I am using older snowflake connector package and I have the same issue.